ios/MLCSwift/Sources/ObjC/LLMChat.mm

```diff
@@ -97,10 +153,97 @@ - (NSString*)runtimeStatsText {
   return [NSString stringWithUTF8String:ret.c_str()];
 }
 
+- (void)processSystemPrompts {
+  process_system_prompts_func_();
+}
+
 - (void)evaluate {
   LOG(INFO) << "Total-mem-budget=" << os_proc_available_memory() / (1 << 20) << "MB";
   llm_chat_->GetFunction("evaluate")();
   LOG(INFO) << "Left-mem-budget=" << os_proc_available_memory() / (1 << 20) << "MB";
 }
 
+- (void)unloadImageMod {
+  image_mod_unload_func_();
+  first_input_after_image = false;
+}
+
+- (void)reloadImageMod:(NSString*)modelLib modelPath:(NSString*)modelPath {
+  std::string lib_prefix = modelLib.UTF8String;
+  std::string model_path = modelPath.UTF8String;
+  std::replace(lib_prefix.begin(), lib_prefix.end(), '-', '_');
+  lib_prefix += '_';
+  Module lib = (*Registry::Get("runtime.SystemLib"))(lib_prefix);
+  image_mod_reload_func_(lib, model_path);
+  first_input_after_image = false;
+}
+
+- (void)resetImageMod {
+  image_mod_reset_func_();
+  first_input_after_image = false;
+}
+
+- (NSString*)runtimeStatsTextImageMod {
+  std::string ret = image_mod_runtime_stats_text_func_();
+  return [NSString stringWithUTF8String:ret.c_str()];
+}
+
+- (void)prefillImage:(UIImage*)image prevPlaceholder:(NSString*)prevPlaceholder postPlaceholder:(NSString*)postPlaceholder {
+  // prefill the previous placeholder string
+  std::string prev_placeholder = prevPlaceholder.UTF8String;
+  prefill_func_(prev_placeholder, false, (int)PlaceInPrompt::kBegin);
+
+  // prefill with image embedding
+  // step 1. get image rawdata: credit from https://stackoverflow.com/a/1262893
+  CGImageRef imageRef = [image CGImage];
+  NSUInteger width = CGImageGetWidth(imageRef);
+  NSUInteger height = CGImageGetHeight(imageRef);
+  CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
+  unsigned char *rawData = (unsigned char*) calloc(height * width * 4, sizeof(unsigned char));
+  NSUInteger bytesPerPixel = 4;
+  NSUInteger bytesPerRow = bytesPerPixel * width;
+  NSUInteger bitsPerComponent = 8;
+  CGContextRef context = CGBitmapContextCreate(rawData, width, height,
+                          bitsPerComponent, bytesPerRow, colorSpace,
+                          kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
+  CGColorSpaceRelease(colorSpace);
+  CGContextDrawImage(context, CGRectMake(0, 0, width, height), imageRef);
+  CGContextRelease(context);
+  // step 2. create rgb array from rgba
+  uint8_t *image_data = (uint8_t*) calloc(height * width * 3, sizeof(uint8_t));
+  int count = 0;
```

Comment:
keep temporal memory as a member so we don't need to re-allocate, add a TODO to move the preproc to the pipeline

Reply:
actually I find the preproc of explicitly casting it to `uint8_t` type necessary here, because I tried to move the part to relax module before and it does not work, probably because of encoding issues.

Reply:
added the whole image conversion (rgba -> rgb, normalization, type conversion) to relax.
also, i think we shall not make `rawData` as a member in the class? because when calling free(rawData) it would delete the variable

Reply:
ok updated